README.md
#fldigi-ruby

===========

## Description

fldigi-ruby is a ruby gem for interfacing with the FLDigi API.  The intent is to be able to use fldigi as a radio modem from ruby scripts (though it's useful for other tasks, as well, such as changing frequencies on a schedule, fetching the current radio frequency for insertion into a log, etc).

I love FLDigi.  One of my favorite features is the API.  But talking directly to the XML-RPC API is a bit of a pain in the ass from random ruby code, so I wrote a gem to abstract it a layer, and remove some of the "ick" factor.  The documenation for the FLDigi API is, to put it kindly, a little lacking.  It basically amounts to a one-line
description of each call (see
[here](http://www.w1hkj.com/FldigiHelp-3.22/xmlrpc-control.html)).

The best documentation has turned out to be reading the FLDigi source.

This gem makes all of this just a bit easier to use.  I've also included a couple of demonstration clients.  The first client allows you to send CQ from the command line, or to send any arbitrary text, using whatever frequency, carrier, mode, and modem you wish.  The second client is a PropNET client written in ruby that uses FLDigi as a modem.  I was using the wrong CRC16 checksum (Did you know there's more than one?  I didn't, either).  I'm now getting valid decodes on the PropNET site, so I'm assuming I've finally got it right.  

This code has been tested on Linux and OS/X, and so far, works perfectly.  I do not own or currently have access to a Windows box, so your mileage may vary there.  Please provide feedback for features and bugs.  My time is limited, but I'll do what I can.  My intent is that library calls won't change, only new calls will be added, but at this point, it's early enough in the development cycle that I can't promise that.  In fact I've already broken that promise at least once.  Turns out I didn't think through the freq() method well when I first started out.  Had to change it from a variable to a method, which would have broken any pre-existing code.  But it had to be done.  You can see what's changed from version to version by viewing the [NEWS](https://github.com/jfrancis42/fldigi-ruby/blob/master/NEWS) file.
    
There's a lot to be desired in things like validating and error-checking of user input.  For example, if you enter bogus fields for phg (for propnet), it'll either send bad data, or barf, depending on which value you hork up.  Input validation is on the very long list of To-Do items.

## First Things First

Before you can have any hope of using this library, you first need to make sure that FLDigi is installed and working correctly and that it's correctly talking to your radio.  If either of these aren't working, you have little to no chance of doing anything useful.

You should be able to have a PSK-31 conversation with FLDigi.  You should have your computer audio wired to your radio, and you should have all of your levels set right (ie, don't overdrive the transmit - if your ALC is moving, you've got it set too high).  If you can have a nice digital QSO with somebody with your setup in it's current condition, you're 1/3 of the way there.

Second step is to make sure that FLDigi can control the mode and frequency of your radio.  You should be able to change bands and frequencies by clicking on the FLDigi user interface.  Obviously, this requires some wiring between your computer and your radio.  With most radios, this is pretty simple.  I bought USB cables on EBay for about $20 each for my Icom and Yaesu radios that take care of this.  It's certainly possible to homebrew your own interfaces (I used to do it that way in Ye Olden Dayes of Yore).  If you can't control the dial frequency, you're stuck with manual tuning, and FLDigi only has about 3khz of spectrum to work with.  Which, in turn, means your code only has about 3khz to work with.  You can certainly still do some things, but you're severely crippling yourself.  If you're good here, that's 2/3 done.

Finally, you need to install the library.  I'm going to make the assumption you've already got ruby installed.  Which means you just need to add the fldigi gem.  On a Mac or Linux, open a terminal, and run:

```bash
sudo gem install fldigi
```

This will go fetch and install the latest version of the gem on your machine.  As the code is still under development, it might be worthwhile to update it regularly (I keep finding bugs and squashing them).  If you have Windows, you'll need to do whatever the equivalent function is to download and install the gem.  I've tested this pretty heavily on Mac and Linux, but I don't own a Windows machine, so I've never tested it in that environment.  I *should* work, but I honestly don't know.  Once you've got the libraries installed, you're 3/3 done with your preparations.

You will also need the Digest CRC gem (digest-crc) written by Hal Brodigan.  This gem is necessary for computing the CCITT CRC16 checksum used for propnet.  Install it like this:

```bash
sudo gem install digest-crc
```

## How to Use

To use the API, you'll first need to require it:

```ruby
require 'fldigi'
```

Next, connect to fldigi.  There are three parameters that can be supplied here.  The simplest is to supply no parameters, which makes the assumption that fldigi is running on the default port on the same host that this script is running on (ie, localhost/127.0.0.1), and that you've configured both FLDigi and your radio for remote control of frequency, mode, etc:

```ruby
fldigi=Fldigi.new()
```

If you have not wired and configured your radio for remote control, do this, instead:

```ruby
fldigi=Fldigi.new(false)
```

You can also optionally supply an IP address and a port number if you're trying to control FLDigi on another host or on a nonstandard port.

The variable 'fldigi' is now your "handle" for talking to FLDigi.  Be aware that FLDigi does some semi-illegal things with XMLRPC, including sometimes returning 'nil' as a result.  This is not uncommon, but technically not allowed by the standard.  So you have to tell ruby to allow for this deviation:

```ruby
XMLRPC::Config::ENABLE_NIL_PARSER=true
```

Most parameters within the object default to something sane at the time of creation, but there are a few things you might want to set.  Be aware that the fldigi gem works by keeping an internal state, then "pushing" that state to the radio when requested.  Though some commands trigger an immediate API call to FLDigi, most manipulate
internal object state, which does not affect the radio until you "sync" them.  Let's set up some state (even though some of these are already defaulted to the values we're setting):

Choose the PSK31 modem.

```ruby
fldigi.modem="BPSK31"
```

Use a carrier frequency of 1khz (the "sweet spot" in most radios).

```ruby
fldigi.carrier=1000
```

Set the dial frequency to 14.070mhz (final xmt freq is dial+carrier, or 14.071mhz).

```ruby
fldigi.dial_freq=14070000
```

Turn on squelch.

```ruby
fldigi.squelch=true
```

Set the squelch value to 3.0 (reasonable for most radios).

```ruby
fldigi.slevel=3.0
```

Turn on AFC.

```ruby
fldigi.afc=true
```

Note that at this point, nothing on the radio has actually changed (nor has anything in the FLDigi program).  All we've done is to prepare the object to sync to FLDigi.  Now let's "push" these changes:

```ruby
fldigi.config()
```

Now things should change.  FLDigi should show the BPSK31 modem selected, the carrier at 1000hz, AFC on, and squelch on and set to 3.0.  The radio should be tuned to 14.070mhz.

Now we can do a little housekeeping before we start sending and receiving data.  We'll make sure we're in receive mode, and clear all the buffers (sent and received) in FLDigi.  Note that these commands execute immediately, without the need to "sync":

```ruby
fldigi.receive()
fldigi.clear_tx_data()
fldigi.get_tx_data()
fldigi.get_rx_data()
```

We're ready to rock and roll.  We can send data, receive data, etc. Let's send a CQ.  There's two ways we can do this.  We could construct a CQ string and send it to FLDigi, or tell the gem what our callsign is and call the cq method.  We'll do the latter.  the cq() method puts the CQ text into the object's output buffer, and send_buffer() sends
this buffer to FLDigi and initiates transmission:

```ruby
fldigi.call="N0CLU"
fldigi.cq()
fldigi.send_buffer()
```

At this point, the radio should click over to transmit, and the CQ call should be sent.  The radio will automatically switch back to receive once the CQ is sent (the send_buffer() method watches FLDigi until all buffered text has been transmitted).

That's it.  You can also send arbitrary text like this:

```ruby
fldigi.add_tx_string("Now is the time for all good men to come to the aid of their country de N0CLU")
fldigi.send_buffer()
```

You can see everything that has been received since the last time you asked like this:

```ruby
puts get_rx_data()
```

Here's a simple example to call CQ where you specify the dial frequency and the desired carrier.  This example will transmit on 14071000hz:

```ruby
require 'fldigi'

  fldigi=Fldigi.new()
  fldigi.call="N0CLU"
  fldigi.carrier=1000
  fldigi.dial_freq=14070000
  fldigi.modem="BPSK31"
  if fldigi.config()
    fldigi.clear_tx_data()
    fldigi.get_tx_data()
    fldigi.cq()
    fldigi.send_buffer()
  end
```

Here's a different way to do the same thing.  This time, we specify the desired transmit frequency, rather than specifying the dial frequency (just set fldigi.carrier to the sweet spot in your audio first).  This example also transmits on 14071000hz, it's just a
different way of doing the same thing:

```ruby
  require 'fldigi'

  fldigi=Fldigi.new()
  fldigi.call="N0CLU"
  fldigi.carrier=1000
  fldigi.freq(14071000)
  fldigi.modem="BPSK31"
  if fldigi.config()
    fldigi.clear_tx_data()
    fldigi.get_tx_data()
    fldigi.cq()
    fldigi.send_buffer()
  end
```

Here's a simple example to send a random bit of text:

```ruby
  require 'fldigi'

  fldigi=Fldigi.new()
  fldigi.dial_freq=18110000
  fldigi.carrier=1200
  fldigi.modem="BPSK63"
  if fldigi.config()
    fldigi.clear_tx_data()
    fldigi.get_tx_data()
    fldigi.add_tx_string("This is a random bit of text de N0CLU")
    fldigi.send_buffer()
  end
```

Here's a simple example to send a PropNet beacon (note that PropNet functionality is untested):

```ruby
  require 'fldigi'

  fldigi=Fldigi.new()
  fldigi.call="N0CLU"
  fldigi.band=20
  fldigi.phg="PHG410015"
  fldigi.grid="CN87wu"
  fldigi.propnet()
  if fldigi.config()
    fldigi.clear_tx_data()
    fldigi.get_tx_data()
    while true
      fldigi.send_buffer()
      sleep fldigi.delay
    end
  end
```

## Included Scripts

### digitool.rb

The digitool.rb script does a bunch of random (and hopefully useful) things.  It'll call CQ for you, it'll randomly tune around the band segment you're in, finding QSOs and printing them, and other assorted things.  Mostly, it serves as an example of how to use the fldigi library.  I tried to make the defaults reasonably sane.  For example, unless you specify otherwise (by using --carrier), the audio carrier defaults to 1000hz.  so if you say "./digitool.rb --txfreq 14071000", it'll set your dial frequency to 14070000hz and your carrier to 1000hz, so your transmit (and receive) frequency will be 14071000hz.  If you tell it "./digitool.rb --dialfreq 14070000", you'll get exactly the same transmit/receive frequency.  Hopefully, being able to specify the dial frequency, the transmit frequency, and/or the carrier frequency will clear up some of the problems that newcomers have to using digital modes.  I sometimes find it useful to do something like:

```
while [ 1 ]; do ./digitool.rb --call n0clu --cq --txfreq 14106500 --carrier 1500 --modem Olivia-32-1K --listen 60; done
```

This will transmit a CQ on 20M in Olivia 32/1000, listen for sixty seconds, then repeat forever.  If/when I get an answer to my CQ, I just ^C the script, and move over to FLDigi and continue the QSO.

### propnet.rb

Connects to a local FLDigi instance with the default API port and sends a PropNet beacon.  Has the option of continuing to do so at a specified interval forever.  Note (see above) that the PropNet client may or may not be working.  I've tried sending beacons with it, and they don't show up on the PropNet page.  That may be bad luck, or it may be broken.  Or more precisely, it may (or may not) need to have it's CRC16 calculation broken to match the (possibly) broken one used by PropNet.  I'm a little unclear on all of this, as some online discussion claims it's broken, and other discussion says it isn't.  At any rate, I couldn't get it to work as-is, though it does follow their documented format precisely.  Making it work is on my to-do list.  Example use:

```
./propnet.rb --call N0CLU --band 40 --phg PHG410015 --grid DM79wu
```

See [the PropNet web site](http://propnet.org) for more information, such as how to set --phg.

### listen.rb

listen.rb and talk.rb are a pair of programs I wrote out of sheer frustration.  I'm constantly wanting to test different propagation, antennas, modes, and power levels between two points, but finding someone to man the station at random odd hours to do the testing has not proven fruitful.  Not to mention, even if I could find a willing victim, it might be a different victim with different antennas, power levels, and equipment each time, so results from one test might not be comparable to results from a second test.  This pair of programs is the result of that frustration.  The idea is simple.  I run listen.rb on my home station.  Every sixty seconds, it changes to the next (out of five) frequency, and listens for sixty seconds for my call sign on the mode I've specified.  If it hears it, it sends me a text message on my phone with the frequency, signal quality, and timestamp.  Meanwhile, I pack up a second station and take it with me wherever I want to test from (camping, a business trip, a city park a few miles down the road, whatever).  I plug in the Signalink, the radio, and the GPS receiver, and then fire up talk.rb.  Assuming I have my clock set accurately on both stations (I used ntpd on the home station and a combination of ntpd and gpsd on the remote station), talk.rb starts sending out location data once per minute, on each of five bands.  Any time my home station hears my remote station, I get a text message that tells me how well it was heard.  Viola, problem solved with technology.  If I'm camping out of range of cell service, I simply read through the FLDigi logs when I return home and correllate the timestamps in the logs with the notes I made in the field (regarding power, antenna, etc).

So, specifics...  listen.rb requires several parameters.  It requires a call sign to listen for using `--call`, a list of five frequencies to step through (or you can use my default list using `--defaults`), a modem (it defaults to BPSK-31), and a method of sending you messages (currently configured for sending texts via Google Voice or PushOver (both free services).  Both Google Voice and PushOver require login credentials, so either or both service credentials must be specified in a YAML file that you specify with `--creds`.  It's a standard YAML file, and looks like this:

```
---
gvlogin: n0clu@gmail.com
gvpasswd: n0treallyap@ssw0rd
pouser: uJDjjd77djUUUdxllvyw8EN9qzzyP
potoken: kAidUen74nahDUYembdye886cjd73nd
```

`gvlogin` is your Google login, `gvpasswd` is your Google password, `pouser` is your PushOver UserID, and `potoken` is your PushOver API token.  You can put creds for both services in the same file if you wish, but it's only required to have one (either one) of them.

### talk.rb

talk.rb has essentially the same options as listen.rb, with a couple of exceptions.  First, it will require you to specify the number of iterations through the five frequencies you'd like to make.  It's not intended to run forever like a beacon, it's intended to run through the set two or three times (to make sure you've got at least two or three chances on each band), then you stop, change something (power, location, antenna, whatever), and try again.  Once it changes to a new band, you have ten seconds before transmissions start.  This is to give you a chance to listen for other activity on frequency that you might interfere with.  If somebody is on your exact frequency, hit "^C" and wait until they're done (or click the carrier over to an empty part of the band in FLDigi using your mouse).  You also either need to be running gpsd on your remote system, or specify your latitude, longitude, and altitude manually.  Latitude and longitude are specified as decimals (ie, 123.456789).  Latitude is positive in the Northern Hemisphere, negative in the Southern.  Longitude is negative in the Western Hemisphere, positive in the Eastern.  Altitude is in feet because I'm an American and because I'm a pilot (both of whom have standardized on the "wrong" system).  There are roughly 3.3ft in one meter.  As with listen.rb, you can optionally specify a list of five frequencies, or use my defaults.  The modem must match that used by listen.rb.

===========

If you look at digitool.rb and propnet.rb, there are working example of all of this and more.  There is also some other random information on the [fldigi-ruby home page](http://fldigi.gritch.org), though I'm slowly moving that information into this document.

===========

73
Jeff
N0GQ


NEWS
Version History:

0.1.15 - 01/04/2015 - jfrancis - Added --getmodem option to
digitool.rb.  Also added a get_modem() method to the object in
preparation for some future functionality (at this point, simply
replaces the API call "modem.get_name").

0.1.14 - 12/19/2014 - jfrancis - Write output to an automatically
named file.  File name is date and time stamp.

0.1.13 - 12/18/2014 - jfrancis - Ok, I admit it.  I'm terrible at
remembering to update this file with new revs of code.  I'll try to be
better.  This version adds the option to digitool to write output to a
file, as well as the screen.

0.1.6 - 08/18/2014 - jfrancis - When the frequency is set, try a
second time if the first set fails. Some radios seem to work fine if
changing frequencies within the same band, but require two tries if
changing bands. If the frequency change fails on the first try, try a
second time, then fail if the second change fails, as well.

0.1.5 - 07/xx/2014 - jfrancis - Did a bunch of general code
clean-up. Looks like a bug in FLDigi related to AFC status
reporting. I can reproduce. Will talk with the FLDigi guys. Added CPS
output to debug data for each batch of data sent.

0.1.4 - 07/28/2014 - jfrancis - Ok, it turns out some XMLRPC libs
return 1 and 0 for true and false, others return true and false.
Changed code to account for both.

0.1.3 - 07/28/2014 - jfrancis - I kind of knew in the back of my mind
that I wasn't handling return status right, but it finally bit me in
the butt on a camping trip this past weekend, so I cleaned it up.
Still more to do, but at least now it properly checks things in
fldigi.config() and properly returns status.  I'm going to further
clean it up so it returns *why* it failed, but at least now it seems
to properly check things.  I also added the fldigi.debug variable,
which, if set to true, will spew debug data to stdout.

0.1.2 - 07/16/2014 - jfrancis - Setting the carrier was broken.  The
code was comparing the current carrier with what the carrier was
previously set to, and only pushing the change if they differed.  This
means that if the user clicked the carrier to a new spot, or of AFC
caused it to drift, updating the carrier had no effect.  Changed to
code to compare the desired carrier with the actual carrier, and make
the change based on that, rather than the value it was previously set
to.

0.1.1 - 07/09/2014 - jfrancis - Crap, I committed the cardinal
developer's sin.  I made a change to the code and did a commit without
testing it first.  After all, what could go wrong?  Radio gear was
still packed from a camping trip, and didn't feel like hooking it up
just to test.  Got a little overzealous with the to_i vs. to_f and
tried to send FLDigi a frequency as an int instead of a float.  Hope
there's not more.  This should be a hint to me that it's time to add
some automated regression testing.  It's on the list...

0.1.0 - 07/02/2014 - jfrancis - Finally added everything on github.
No real code changes.  Finally ready for people to start using for
something useful.

0.0.15 - 07/02/2014 - jfrancis - Oops. No changes, just messed up
the metadata.

0.0.14 - 07/02/2014 - jfrancis - Be more consistent about to_f vs.
to_i.  Also, add a function to return actual radio freq (instead of
specified freq).

0.0.13 - 06/11/2014 - jfrancis - A bunch of clean-up here.  I almost
certainly broke some code with this one.  I changed @freq to
@dial_freq to more accurately reflect what's going on, and added a
method self.freq() to get/set the actual transmit frequency.
Anything that did fldigi.freq=1234567 before will need to change to
fldigi.dial_freq=1234567.  My apologies, but it just needed to be
done.

0.0.12 - 06/05/2014 - jfrancis - Allows for additional data to be
appended to the output queue during transmit.  Also now defaults to
not displaying a running output of transmitted data (though it can
be switched back on with a 'true' flag to send_buffer()).

0.0.11 - 05/29/2014 - jfrancis - Added optional offset.

0.0.10 - 05/26/2014 - jfrancis - Forgot the 'require'.

0.0.9 - 05/26/2014 - jfrancis - Added a lot of code comments.

0.0.8 - 05/26/2014 - jfrancis - Fixed 10m propnet frequency. Got
carried away with 0.0.7 fix, and had one too many zeros in there.

0.0.7 - 05/26/2014 - jfrancis - All propnet frequencies were low by
an order of magnitude (ie, left off a zero).

0.0.6 - 05/25/2014 - jfrancis - First public release on
rubygems.org.


talk.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

# Jeff Francis, N0GQ, jeff@gritch.org
#
# See http://fldigi.gritch.org for documentation.

require 'io/console'
require 'time'
require 'json'
require 'thread'
require 'rubygems'
require 'trollop'
require 'fldigi'

$my_lat=0
$my_lon=0
$my_alt=0
$my_last=0
$my_new=false
$gpsd_lock=false

def gpsd(host, port)
  s=TCPSocket.open(host, port)
  s.puts '?WATCH={"enable":true,"json":true}'

  while true
    s.puts '?POLL;'
    
    line=s.gets
    now=Time.now
    updated=false
    thing=JSON.parse(line)

    tmp_lat=thing['lat'].to_f
    tmp_lon=thing['lon'].to_f
    tmp_alt=thing['alt'].to_f
    
    if tmp_lat!=0
      $my_lat=tmp_lat
      updated=true
    end

    if tmp_lon!=0
      $my_lon=tmp_lon
      updated=true
    end

    if tmp_alt!=0
      $my_alt=(tmp_alt*3.281).to_i
      updated=true
    end

    if updated==true
      $my_last=now
      $my_new=true
    end

    if tmp_lat+tmp_lon+tmp_alt!=0.0 and !$gpsd_lock
      $gpsd_lock=true
    end

    sleep 1
  end
end

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do command line option parsing.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

opts=Trollop::options do
  opt :debug, "Show extra debug info (not useful for most users)"
  opt :host, "The FLDigi host (defaults to localhost)", :type => :string
  opt :port, "The FLDigi port (defaults to 7362)", :type => :string
  opt :modem, "Modem (defaults to BPSK31)", :type => :string
  opt :modems, "List available FLDigi modems"
  opt :call, "The call sign to watch for", :type => :string
  opt :iterations, "How many transmit iterations through all five frequencies", :type => :string
  opt :defaults, "Use default freqs of 3581000, 7081000, 10143000, 14071000, 28121000"
  opt :freq1, "First frequency to listen on", :type => :string
  opt :freq2, "Second frequency to listen on", :type => :string
  opt :freq3, "Third frequency to listen on", :type => :string
  opt :freq4, "Fourth frequency to listen on", :type => :string
  opt :freq5, "Fifth frequency to listen on", :type => :string
  opt :gpsd, "Use gpsd for location"
  opt :gpsdhost, "Specify gpsd host (defaults to localhost)", :type => :string
  opt :gpsdport, "Specify gpsd port (defaults to 2947)", :type => :string
  opt :lat, "Manually specify latitude (dd.ddddd)", :type => :string
  opt :lon, "Manually specify longitude (dd.ddddd)", :type => :string
  opt :alt, "Manually specify altitude (feet)", :type => :string
end

if opts[:iterations_given]
  iterations=opts[:iterations].to_i*5
else
  iterations=5
end

if opts[:gpsd_given]
  if opts[:gpsdhost_given]
    gpsd_host=opts[:gpsdhost]
  else
    gpsd_host="127.0.0.1"
  end

  if opts[:gpsdport_given]
    gpsd_port=opts[:gpsdport].to_i
  else
    gpsd_host=2947
  end
else
  if !opts[:lat_given] or !opts[:lon_given] or !opts[:alt_given]
    puts "Must specify location manually if not using --gpsd"
    exit
  end

  manual_lat=opts[:lat].to_f
  manual_lon=opts[:lon].to_f
  manual_alt=opts[:alt].to_i
end

if !opts[:defaults_given]
  if !opts[:freq1_given] or !opts[:freq2_given] or !opts[:freq3_given] or !opts[:freq4_given] or !opts[:freq5_given]
    puts "Must specify either --defaults or all five RX frequencies (duplicates are allowed)"
    exit
  else
    freq1=opts[:freq1].to_i
    freq2=opts[:freq2].to_i
    freq3=opts[:freq3].to_i
    freq4=opts[:freq4].to_i
    freq5=opts[:freq5].to_i
  end
end

if opts[:call_given]
  call=opts[:call].downcase
else
  puts "--call is required"
  exit
end

if opts[:host_given]
  host=opts[:host]
end

if opts[:port_given]
  port=opts[:port].to_i
end

if opts[:modems_given]
  puts fldigi.list_modems()
  exit
end

if opts[:modem_given]
  fldigi.modem=opts[:modem]
end

if opts[:gpsd_given]
  # Talk to gpsd.
  gps=Thread.new { gpsd(gpsd_host, gpsd_port) }
  gps.abort_on_exception=true

  while !$gpsd_lock
    puts "Waiting for GPS lock..."
    sleep 1
  end
  puts "GPS lock achieved."
end

# Talk to the mother ship.
fldigi=Fldigi.new()

# May not need this. Seems to depend on ruby version.
if !XMLRPC::Config::ENABLE_NIL_PARSER
  v, $VERBOSE=$VERBOSE, nil
  XMLRPC::Config::ENABLE_NIL_PARSER=true
  $VERBOSE=v
end

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do the needful.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

# Set all of the random things that need setting.
fldigi.receive
fldigi.spot=false
fldigi.carrier=1000
fldigi.afc=false
fldigi.clear_tx_data()
fldigi.get_tx_data()
fldigi.get_rx_data()
fldigi.config()

# Go into an infinite talk loop.
if opts[:defaults_given]
  f=[3581000, 7081000, 10143000, 14071000, 28121000]
else
  f=[freq1, freq2, freq3, freq4, freq5]
end
search=true
count=0
while count < iterations
  # Set the tx freq based on the time.
  t=Time.now()
  m=t.to_s.split[1].split(':')[1][1].to_i
  s=t.to_s.split[1].split(':')[2].to_i
  if s>=0 and s<=1
    if m>4
      m=m-5
    end

    fldigi.freq(f[m])
    fldigi.config()

  elsif s>=10 and s<=15

    # Send a location string on the new band.
    lat=$my_lat.to_s
    lon=$my_long.to_s
    alt=$my_alt.to_s

    puts "#{t.to_s} #{fldigi.freq()}hz #{$my_lat} #{$my_lon} #{$my_alt}ft"

    if opts[:gpsd_given]
      fldigi.add_tx_string("test location test location test location #{$my_lat}:#{$my_lon}:#{$my_alt}ft de #{opts[:call]} #{opts[:call]} #{opts[:call]}")
    else
      fldigi.add_tx_string("test location test location test location #{manual_lat}:#{manual_lon}:#{manual_alt}ft de #{opts[:call]} #{opts[:call]} #{opts[:call]}")
    end
    fldigi.send_buffer(false)
    count=count+1
  else
    sleep 1
  end
end


propnet.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

# Jeff Francis, N0GQ, jeff@gritch.org
#
# Connects to a local FLDigi instance with the default API port.  Example use:
#
# ./propnet.rb --call N0CLU --band 40 --phg PHG410015 --grid DM79wu
#
# See http://propnet.org for information, such as how to set --phg.

require "xmlrpc/client"
require 'zlib'
require 'base64'
require 'time'
require 'rubygems'
require 'trollop'
require 'fldigi'

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do command line option parsing.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

opts=Trollop::options do
  opt :call, "Call sign", :type => :string
  opt :host, "The FLDigi host (defaults to localhost)", :type => :string
  opt :port, "The FLDigi port (defaults to 7362)", :type => :string
  opt :psk63, "Use PSK63 (defaults to PSK31)"
  opt :band, "Band to beacon in (80, 40, 20, 17, 15, 12, 10, 6)", :type => :string
  opt :grid, "Grid square for use in PropNet (6 characters)", :type => :string
  opt :phg, "PHG for PropNet (format PHG410015 - beacons/hour will be extracted from this)", :type => :string
end

fldigi=Fldigi.new()

# May not need this. Seems to depend on ruby version.
if !XMLRPC::Config::ENABLE_NIL_PARSER
  v, $VERBOSE=$VERBOSE, nil
  XMLRPC::Config::ENABLE_NIL_PARSER=true
  $VERBOSE=v
end

if opts[:psk63_given]
  fldigi.modem="BPSK63"
else
  fldigi.modem="BPSK31"
end

if opts[:call_given]
  fldigi.call=opts[:call].downcase
else
  puts "Must specify --call"
  exit
end

if opts[:band_given]
  fldigi.band=opts[:band]
else
  puts "Must specify --band"
  exit
end

delay=nil
if opts[:phg_given]
  if opts[:phg][0,3]!="PHG"
    puts "--phg must begin with the letters PHG"
    exit
  else
    fldigi.phg=opts[:phg]
  end
else
  puts "Must specify --phg"
  exit
end

if opts[:band_given]
else
  puts "Must specify --band"
  exit
end

grid=nil
if opts[:grid_given]
  fldigi.grid=opts[:grid]
  if opts[:grid].length!=6
    puts "--grid must be six characters"
    exit
  end
else
  puts "Must specify --grid"
  exit
end

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do the needful.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

# Set all of the random things that need setting.
fldigi.receive
fldigi.clear_tx_data()
fldigi.get_tx_data()
fldigi.get_rx_data()
fldigi.propnet_config()
fldigi.config()

if fldigi.delay
  while true
    puts "#{Time.now().to_s} Sending beacon..."
    fldigi.propnet()
    fldigi.send_buffer()
    puts "#{Time.now().to_s} Done.\n\n"
    sleep fldigi.delay
  end
else
  fldigi.propnet()
  fldigi.send_buffer()
end


listen.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

# Jeff Francis, N0GQ, jeff@gritch.org
#
# See http://fldigi.gritch.org for documentation.

require 'io/console'
require 'time'
require 'yaml'
require 'rubygems'
require 'trollop'
require 'fldigi'

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do command line option parsing.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

opts=Trollop::options do
  opt :debug, "Show extra debug info (not useful for most users)"
  opt :host, "The FLDigi host (defaults to localhost)", :type => :string
  opt :port, "The FLDigi port (defaults to 7362)", :type => :string
  opt :call, "The call sign to watch for", :type => :string
  opt :modem, "Modem (defaults to BPSK31)", :type => :string
  opt :modems, "List available FLDigi modems"
  opt :creds, "YAML file containing Google Voice/PushOver credentials", :type => :string
  opt :gvoice, "Send a text to this number using Google Voice upon call match", :type => :string
  opt :pushover, "Send a message via PushOver upon call match"
  opt :defaults, "Use default freqs of 3581000, 7081000, 10143000, 14071000, 28121000"
  opt :freq1, "First frequency to listen on", :type => :string
  opt :freq2, "Second frequency to listen on", :type => :string
  opt :freq3, "Third frequency to listen on", :type => :string
  opt :freq4, "Fourth frequency to listen on", :type => :string
  opt :freq5, "Fifth frequency to listen on", :type => :string
end

if !opts[:defaults_given]
  if !opts[:freq1_given] or !opts[:freq2_given] or !opts[:freq3_given] or !opts[:freq4_given] or !opts[:freq5_given]
    puts "Must specify either --defaults or all five RX frequencies (duplicates are allowed)"
    exit
  else
    freq1=opts[:freq1].to_i
    freq2=opts[:freq2].to_i
    freq3=opts[:freq3].to_i
    freq4=opts[:freq4].to_i
    freq5=opts[:freq5].to_i
  end
end

if opts[:creds_given]
  creds=YAML.load(File.read(opts[:creds]))
  
  if opts[:gvoice_given] and creds.has_key?("gvlogin") and creds.has_key?("gvpasswd")
    require 'googlevoiceapi'
    gvapi=GoogleVoice::Api.new(creds['gvlogin'],creds['gvpasswd'])
  end
  
  if opts[:pushover_given] and creds.has_key?("pouser") and creds.has_key?("potoken")
    require 'pushover'
  end  
else
  puts "Invalid notification method (either you forgot --gvoice or --pushover, or your --creds are broken)."
  exit
end

if opts[:host_given]
  host=opts[:host]
end

if opts[:port_given]
  port=opts[:port].to_i
end

if opts[:call_given]
  call=opts[:call].downcase
else
  puts "--call is required"
  exit
end

# Talk to the mother ship.
fldigi=Fldigi.new()

# May not need this. Seems to depend on ruby version.
if !XMLRPC::Config::ENABLE_NIL_PARSER
  v, $VERBOSE=$VERBOSE, nil
  XMLRPC::Config::ENABLE_NIL_PARSER=true
  $VERBOSE=v
end

# Give the user a list of supported modems that can be specified with
# the '--modem' parameter.
if opts[:modems_given]
  puts fldigi.list_modems()
  exit
end

if opts[:modem_given]
  fldigi.modem=opts[:modem]
end

# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
# Do the needful.
# -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

# Set all of the random things that need setting.
fldigi.receive
fldigi.spot=true
fldigi.clear_tx_data()
fldigi.get_tx_data()
fldigi.get_rx_data()
fldigi.config()

# Go into an infinite listen loop.
if opts[:defaults_given]
  f=[3581000, 7081000, 10143000, 14071000, 28121000]
else
  f=[freq1, freq2, freq3, freq4, freq5]
end
search=true
last_freq=0
while true
  # Set the rx freq based on the time.
  t=Time.now()
  m=t.to_s.split[1].split(':')[1][1].to_i
  if m>4
    m=m-5
  end

  if last_freq!=f[m]
    last_freq=f[m]
    fldigi.freq(f[m])
    fldigi.carrier=1000
    fldigi.config()
    puts "#{t.to_s} #{fldigi.freq()}hz"
  end

  quality=fldigi.quality().to_i
  if quality<25
    if rand(2)==0
      fldigi.search_up()
    else
      fldigi.search_down()
    end
  else
    tmp="(#{quality} #{fldigi.freq()}hz #{Time.now().to_s}) #{fldigi.get_rx_data()}"
    puts tmp
    if tmp.downcase.match(call)
      if opts[:gvoice_given]
        gvapi.sms(opts[:gvoice],tmp)
      end

      if opts[:pushover_given]
        Pushover.notification(message: ARGV[0], title: opts[:call],
                              user: creds['pouser'] , token: creds['potoken'])
      end
    end
  end
  sleep 3
end
